test/e2e/clusterctl_upgrade.go

```diff
@@ -65,10 +65,12 @@ type ClusterctlUpgradeSpecInput struct {
 	// provider contract to use to initialise the secondary management cluster, e.g. `v1alpha3`
 	InitWithProvidersContract string
 	SkipCleanup               bool
+	PreCreateCluster          func(managementClusterProxy framework.ClusterProxy)
```

Comment:
What is the semantic of this func? (We create multiple workload clusters in this test)

Reply:
For instance, CAPM3 provider requires BMH CRs available in the same namespace before creating the cluster. Without this hook there no way to create those CRs.

Reply:
Am I right to assume we also need this later in our 2nd management cluster?

Reply:
True, /hold I will add that

Reply:
I'm also not sure if the current location works. Given it's outside of `It`. When will ginkgo execute it?

If that's something that CAPM3 always needs I'm also not sure if that should be only a per-test case.
It might fit better into the bootstrapping (where we deploy all providers, ...)

Reply:
Thanks for the suggestion I have made some changes yet I would like to hold this PR check if it will be feasible to add those crs with the template

Reply:
Sure no problem
/hold

Reply:
Using the BMH in the template can help but in the concept of metal3, baremetal hosts should already be available in the namespace and not a part of the template because they are more static resources representing physical layer. So I prefer setting them through the hooks with some conditioning to check before which cluster the hook was called

Reply:
/unhold